### PR TITLE
Argument Relevance for Synthesis Conjectures

### PR DESCRIPTION
**Pull request by ajreynol**
EDIT : merged pull request #1323, to take care of part of this pull request.

This introduces a new algorithm for detecting when argument positions of functions-to-synthesize are irrelevant.  It is a working version of a solution for #1210.

The algorithm is implemented in the CegProcessConj class, which is invoked when a synthesis conjecture is registered.  Based on an analysis, the grammar construction algorithm in CegGrammarCons can restrict the grammar for the function-to-synthesize.

TODO:
- Symmetry breaking for user-provided grammars,
- Generalize the cases covered by the algorithm.

### CODE REVIEW

**Comment by ajreynol** [comment]
I've taken some of this PR and added it to #1323 to expedite getting the changes into master that are needed for Haniel to get started working, so let's merge #1323 before this one.

File: src/theory/quantifiers/sygus_process_conj.cpp [hunk 1]
```diff
@@ -27,11 +27,546 @@ namespace CVC4 {
...
+      else
+      {
+        std::unordered_map<Node,
+                           std::unordered_set<Node, NodeHashFunction>,
+                           NodeHashFunction>::iterator itf = free_vars.find(nn);
+        Assert(itf != free_vars.end());
+        for (std::unordered_set<Node, NodeHashFunction>::iterator itfv =
+                 itf->second.begin();
+             itfv != itf->second.end();
+             ++itfv)
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.cpp]: This is again a case where a for each loop would be more compact and easier to read.

File: src/theory/quantifiers/sygus_process_conj.cpp [hunk 2]
```diff
@@ -27,11 +27,546 @@ namespace CVC4 {
...
+    }
+    base = base[0][1];
+  }
+  std::vector<Node> conjuncts;
+  getComponentVector(AND, base, conjuncts);
+
+  // process the conjunctions
+  for (std::map<Node, CegConjectureProcessFun>::iterator it = d_sf_info.begin();
+       it != d_sf_info.end();
+       ++it)
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.cpp]: Ditto (for each).

File: src/theory/quantifiers/sygus_process_conj.cpp [hunk 3]
```diff
@@ -47,51 +582,223 @@ void CegConjectureProcess::initialize(Node n, std::vector<Node>& candidates)
...
+
+bool CegConjectureProcess::isArgRelevant(Node f, unsigned i)
+{
+  std::map<Node, CegConjectureProcessFun>::iterator its = d_sf_info.find(f);
+  if (its != d_sf_info.end())
   {
-    base = n[0][1];
+    return its->second.isArgRelevant(i);
   }
   else
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.cpp]: Remove `else`?

File: src/theory/quantifiers/sygus_process_conj.cpp [hunk 4]
```diff
@@ -47,51 +582,223 @@ void CegConjectureProcess::initialize(Node n, std::vector<Node>& candidates)
...
+  if (its != d_sf_info.end())
   {
-    for (unsigned i = 0; i < base.getNumChildren(); i++)
-    {
-      conj.push_back(base[i]);
-    }
+    its->second.getIrrelevantArgs(args);
+    return true;
   }
   else
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.cpp]: Ditto.

File: src/theory/quantifiers/sygus_process_conj.h [hunk 5]
```diff
@@ -15,26 +15,234 @@
...
 
+/** This file contains techniques that compute
+ * argument relevancy for synthesis functions
+ *
+ * Let F be a synthesis conjecture of the form:
+ *   exists f. forall X. P( f, X )
+ *
+ * The classes below compute whether certain arguments of
+ * the function-to-synthesize f are irrelevant.
+ * Assume that f is binary function, where possible solutions
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.h]: Typo: ... is **a** binary ...

File: src/theory/quantifiers/sygus_process_conj.h [hunk 6]
```diff
@@ -15,26 +15,234 @@
...
+   *   checkMatch( x1+1, t+1, n_arg_map ) returns true,
+   *   checkMatch( 0, 0, n_arg_map ) returns true,
+   *   checkMatch( x1+1, s, n_arg_map ) returns false.
+   */
+  bool checkMatch(Node cn,
+                  Node n,
+                  std::unordered_map<unsigned, Node>& n_arg_map);
+  /** infer definition
+   *
+   * In the following, we say a term is a "emplate
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.h]: Typo: ... "**t**emplate definition" ...

File: src/theory/quantifiers/sygus_process_conj.h [hunk 7]
```diff
@@ -15,26 +15,234 @@
...
+   * definition" if its free variables
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.h]: Remove newline.

File: src/theory/quantifiers/sygus_process_conj.h [hunk 8]
```diff
@@ -15,26 +15,234 @@
...
+      std::unordered_map<Node, unsigned, NodeHashFunction>& term_to_arg_carry,
+      std::unordered_map<Node,
+                         std::unordered_set<Node, NodeHashFunction>,
+                         NodeHashFunction>& free_vars);
+  /** Assign relevant definition
+   *
+   * If def is non-null,
+   * this function assigns def as a template definition
+   * for the (currently irrelevant) arguments that
+   * occur in a term of the form f(
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.h]: There seems to be something missing: ... of the form f( ?

File: src/theory/quantifiers/sygus_process_conj.h [hunk 9]
```diff
@@ -15,26 +15,234 @@
...
+   *
+   * If def is null, then an argument is marked
+   * as relevant.
+   *
+   * Returns an value rid such that:
```
Comment (mpreiner) [inline src/theory/quantifiers/sygus_process_conj.h]: Typo: ... **a** value ...

File: src/theory/quantifiers/sygus_process_conj.h (hunk 8 above)
Comment (ajreynol) [reply src/theory/quantifiers/sygus_process_conj.h]: Yes, I'll fix.

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit.  I'd still like to hold off on for -> for each unless there is a consensus decision in the style guidelines.

File: src/theory/quantifiers/ce_guided_pbe.h [hunk 10]
```diff
@@ -192,8 +190,7 @@ class CegConjecturePbe {
   Node d_true;
   Node d_false;
   /** parent conjecture
-  * This is the data structure that contains global information about the
-  * conjecture.
+  * This contains global information about the synthesis conjecture.
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/ce_guided_pbe.h]: This comment was out of place, I've replaced it.

**Pull request merged**